### PR TITLE
Convert WaitSet to use C++ Classes

### PR DESCRIPTION
**Pull request by ahcorde**
Related to https://github.com/ros2/rclpy/issues/665

This converts the WaitSet functions to a C++ class. 

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rclpy/src/rclpy/wait_set.cpp [hunk 1]
```diff
@@ -318,23 +283,62 @@ get_ready_entities(const std::string & entity_type, py::capsule pywait_set)
...
-    ret = rcl_wait(wait_set, timeout);
+    ret = rcl_wait(rcl_wait_set_.get(), timeout);
   }
 
-  if (ret != RCL_RET_OK && ret != RCL_RET_TIMEOUT) {
+  if (RCL_RET_OK != ret && RCL_RET_TIMEOUT != ret) {
     throw RCLError("failed to wait on wait set");
   }
 }
+void define_waitset(py::object module)
```
Comment (hidmic) [inline rclpy/src/rclpy/wait_set.cpp:300]: @ahcorde nit: missing a blank line?

File: rclpy/src/rclpy/wait_set.cpp [hunk 2]
```diff
@@ -318,23 +283,62 @@ get_ready_entities(const std::string & entity_type, py::capsule pywait_set)
...
+    "Add a subcription to the wait set structure")
+  .def(
+    "wait_set_add_client", &WaitSet::wait_set_add_client,
+    "Add a client to the wait set structure")
+  .def(
+    "wait_set_add_timer", &WaitSet::wait_set_add_timer,
+    "Add a timer to the wait set structure")
+  .def(
+    "wait_set_clear_entities", &WaitSet::wait_set_clear_entities,
+    "Add a client to the wait set structure")
```
Comment (hidmic) [inline rclpy/src/rclpy/wait_set.cpp:330]: @ahcorde needs an update?

File: rclpy/src/rclpy/wait_set.cpp [hunk 3]
```diff
@@ -318,23 +283,62 @@ get_ready_entities(const std::string & entity_type, py::capsule pywait_set)
...
+  .def(
+    "wait_set_add_event", &WaitSet::wait_set_add_event,
+    "Add an event to the wait set structure")
+  .def(
+    "wait_set_is_ready", &WaitSet::wait_set_is_ready,
+    "Check if an entity in the wait set is ready by its index")
+  .def(
+    "get_ready_entities", &WaitSet::get_ready_entities,
+    "Get list of non-null entities in wait set")
```
Comment (hidmic) [inline rclpy/src/rclpy/wait_set.cpp:339]: @ahcorde nit: perhaps
```suggestion
    "Get list of entities ready by entity type")
```
would be more accurate?

File: rclpy/src/rclpy/wait_set.hpp [hunk 4]
```diff
@@ -30,149 +31,151 @@ namespace py = pybind11;
...
+public:
+  /// Initialize a wait set
+  WaitSet();
+
+  /// Initialize a wait set
+  /**
+   * Raises RCLError if the wait set could not be initialized
+   *
+   * \param[in] node_name string name of the node to be created
+   * \param[in] number_of_subscriptions int
```
Comment (hidmic) [inline rclpy/src/rclpy/wait_set.hpp:45]: @ahcorde nit: 
```suggestion
   * \param[in] number_of_subscriptions a positive number or zero
```
Same elsewhere.

**Review by hidmic — APPROVED** [review]
Overall LGTM, pending green CI

**Comment by ahcorde** [comment]
compiling up-to `rclpy` and testing `rclpy`

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14376/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9152/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12052/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14475/)

**Pull request merged**